From f33620344f61bb10238f37b56df690c8034b9a37 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tim=20R=C3=BChsen?= <tim.ruehsen@gmx.de>
Date: Sat, 29 Feb 2020 18:46:45 +0100
Subject: [PATCH 43/64] talk: Fix uninitialized variable 'nready'

---
 talk/ctl_transact.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/talk/ctl_transact.c b/talk/ctl_transact.c
index 843bdff6..26d4c34d 100644
--- a/talk/ctl_transact.c
+++ b/talk/ctl_transact.c
@@ -70,7 +70,7 @@
 int
 ctl_transact (struct in_addr target, CTL_MSG msg, int type, CTL_RESPONSE * rp)
 {
-  int nready, cc;
+  int nready = 0, cc;
   fd_set read_mask, ctl_mask;
   struct timeval wait;
 
-- 
2.28.0.163.g6104cc2f0b6

