From 7c7d74d528daf6075110b0a782e06f0629ab76ba Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tim=20R=C3=BChsen?= <tim.ruehsen@gmx.de>
Date: Sat, 29 Feb 2020 19:01:40 +0100
Subject: [PATCH 45/64] ftp: Fix strncpy misuse (leading to buffer overflow)

---
 ftp/ftp.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/ftp/ftp.c b/ftp/ftp.c
index 604d3e55..131f914d 100644
--- a/ftp/ftp.c
+++ b/ftp/ftp.c
@@ -175,9 +175,10 @@ hookup (char *host, int port)
     }
 
   if (res->ai_canonname)
-    strncpy (hostnamebuf, res->ai_canonname, sizeof (hostnamebuf));
+    strncpy (hostnamebuf, res->ai_canonname, sizeof (hostnamebuf) - 1);
   else
-    strncpy (hostnamebuf, rhost, sizeof (hostnamebuf));
+    strncpy (hostnamebuf, rhost, sizeof (hostnamebuf) - 1);
+  hostnamebuf[sizeof (hostnamebuf) - 1] = 0;
 
   hostname = hostnamebuf;
   free (rhost);
-- 
2.28.0.163.g6104cc2f0b6

