From 98b85a2fd63462665a0b51322389852c7a8af73d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tim=20R=C3=BChsen?= <tim.ruehsen@gmx.de>
Date: Sat, 29 Feb 2020 19:42:00 +0100
Subject: [PATCH 53/64] telnet/commands.c: Use strdup instead malloc+strcpy

---
 telnet/commands.c | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/telnet/commands.c b/telnet/commands.c
index cd0c75ab..5003ee07 100644
--- a/telnet/commands.c
+++ b/telnet/commands.c
@@ -2585,10 +2585,8 @@ tn (int argc, char *argv[])
     }
 
   free (hostname);
-  hostname = malloc (strlen (hostp) + 1);
-  if (hostname)
-    strcpy (hostname, hostp);
-  else
+  hostname = strdup (hostp);
+  if (!hostname)
     {
       printf ("Can't allocate memory to copy hostname\n");
       return 0;
-- 
2.28.0.163.g6104cc2f0b6

