Index: inetutils-1.5/configure.ac
===================================================================
--- inetutils-1.5.orig/configure.ac	2006-10-23 07:55:10.000000000 +0300
+++ inetutils-1.5/configure.ac	2006-10-23 07:55:23.000000000 +0300
@@ -794,5 +794,5 @@ ifconfig/Makefile
 ifconfig/system/Makefile
 confpaths.h:confpaths.h.in
 headers/Makefile
-doc/Makefile doc/rfc/Makefile])
+])
 AC_OUTPUT
Index: inetutils-1.5/Makefile.am
===================================================================
--- inetutils-1.5.orig/Makefile.am	2006-10-23 07:55:27.000000000 +0300
+++ inetutils-1.5/Makefile.am	2006-10-23 07:55:36.000000000 +0300
@@ -25,7 +25,7 @@ EXTRA_DIST = README-alpha paths ChangeLo
 SUBDIRS = lib headers libinetutils libtelnet glob \
 	inetd telnetd libls ftpd rshd rlogind uucpd rexecd syslogd tftpd \
 	talkd telnet ftp rsh rcp rlogin tftp logger gwhois talk \
-	libicmp ping doc ifconfig
+	libicmp ping ifconfig
 
 DISTCLEANFILES = pathdefs.make paths.defs include/*.h include/*/*.h \
 	include/stamp-h
