#DPATCHLEVEL=0
Index: configure.ac
===================================================================
--- configure.ac.orig	2006-04-20 12:53:13.000000000 +0300
+++ configure.ac	2006-10-03 07:53:32.000000000 +0300
@@ -131,12 +131,12 @@ AC_PROG_YACC
 gl_INIT
 
 ### Checks for libraries.
-AC_CHECK_LIB(bsd, revoke)
+#AC_CHECK_LIB(bsd, revoke)
 AC_CHECK_LIB(nsl, inet_ntoa)
 AC_CHECK_LIB(resolv, gethostname)
 AC_CHECK_LIB(socket, getpeername)
 AC_CHECK_LIB(opie, opiegenerator)
-AC_CHECK_LIB(z, gzdopen)
+#AC_CHECK_LIB(z, gzdopen)
 
 # See if a termcap library is available (under one of several names)
 IU_LIB_TERMCAP
