Index: b/configure.ac
===================================================================
--- a/configure.ac
+++ b/configure.ac
@@ -531,7 +531,7 @@ if test "$ac_cv_header_utmpx_h" = yes; t
   IU_CHECK_MEMBERS([struct utmpx.ut_tv], , , [#include <utmpx.h>])
 fi
 
-IU_CHECK_MEMBERS([stuct ifreq.ifr_index,
+IU_CHECK_MEMBERS([struct ifreq.ifr_index,
                   struct ifreq.ifr_netmask,
                   struct ifreq.ifr_broadaddr], , ,
                  [#include <sys/types.h>
