Index: b/rsh/rsh.c
===================================================================
--- a/rsh/rsh.c
+++ b/rsh/rsh.c
@@ -70,6 +70,8 @@
 # include <sys/select.h>
 #endif
 
+#include "xalloc.h"
+
 #ifdef SHISHI
 #define REALM_SZ 1040
 #endif
Index: b/telnet/authenc.c
===================================================================
--- a/telnet/authenc.c
+++ b/telnet/authenc.c
@@ -40,6 +40,7 @@ static char sccsid[] = "@(#)authenc.c	8.
 #include <arpa/telnet.h>
 #include <libtelnet/encrypt.h>
 #include <libtelnet/misc.h>
+#include <unistd.h>
 
 #include "general.h"
 #include "ring.h"
