From 81676d4013791d32315d098541886e3207b6077d Mon Sep 17 00:00:00 2001
From: Simon Josefsson <simon@josefsson.org>
Date: Sat, 2 May 2009 06:38:27 +0000
Subject: [PATCH] acinclude.m4: Avoid comment to fix script.

---
 acinclude.m4 |    4 ++--
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/acinclude.m4 b/acinclude.m4
index ebb8ee1..6206545 100644
--- a/acinclude.m4
+++ b/acinclude.m4
@@ -1,6 +1,6 @@
 dnl Autoconf macros used by inetutils
 dnl
-dnl Copyright (C) 1996, 1997, 1998, 2002, 2004, 2005, 2007 Free Software Foundation, Inc.
+dnl Copyright (C) 1996, 1997, 1998, 2002, 2004, 2005, 2007, 2009 Free Software Foundation, Inc.
 dnl
 dnl Mostly written by Miles Bader <miles@gnu.ai.mit.edu>
 dnl
@@ -577,7 +577,7 @@ if (sizeof(ac_aggr.m4_bpatsubst([$1], [^[^.]*\.])))
 return 0;])],
                 [AS_VAR_SET(ac_Member, yes)],
                 [AS_VAR_SET(ac_Member, no)])])
-AS_IF([test AS_VAR_GET(ac_Member) = yes], [$2], [$3])dnl
+AS_IF([test AS_VAR_GET(ac_Member) = yes], [$2], [$3])
 AS_VAR_POPDEF([ac_Member])dnl
 ])dnl IU_CHECK_MEMBER
 
-- 
1.6.3.3

